python/pyarrow/_dataset.pyx

```diff
@@ -385,6 +416,32 @@ cdef class Dataset(_Weakrefable):
         """The common schema of the full Dataset"""
         return pyarrow_wrap_schema(self.dataset.schema())
 
+    def filter(self, expression):
+        """
+        Apply a row filter to the dataset.
+
+        Parameters
+        ----------
+        expression : Expression
+            The filter that should be applied to the dataset.
+
+        Returns
+        -------
+        Dataset
+        """
+        cdef:
+            Dataset filtered_dataset
+
+        new_filter = expression
+        current_filter = self._scan_options.get("filter")
+        if current_filter is not None and new_filter is not None:
+            new_filter = current_filter & new_filter
```

Comment:
If you check that `new_filter is not None`, that means we are covering the case that `new_filter` would be None, but in that case we have to set `new_filter = current_filter` ?

Reply:
Passing `None` as the filter is meant to be a way to remove all filters.
To purpose is to allow retrieving fragments of a filtered dataset by doing `result.filter(None).get_fragments()`

Reply:
Ah, that was not clear to me (and should also be documented then). 

Personally I would also not expect such behaviour, since that seems to go against my expectation for composable method chains (based on my experience with pandas, and from a quick check also not what dplyr filter does when passing no filter), where each method does something in addition to the previous one, and not undo it.

Reply:
But I don't see what `.filter(None)` could mean if not to unset the current filter. Would you prefer a dedicated `Dataset.reset_filter()` method?

Reply:
> But I don't see what .filter(None) could mean if not to unset the current filter.

For me the simplest interpretation would be to "not add any new filter". 
Users don't necessarily have the concept of "current filter" in their mind. The dataset is already filtered at that point, and then `filter(None)` would just not filter it any further.



> Would you prefer a dedicated `Dataset.reset_filter()` method?

We could also defer that until there are some actual use cases / requests for this? I am not sure if there is a good use case to first filter and then reset it to be able to call get_fragments (except while interactively exploring, but in such a case you can also easily recreate the object)

Reply:
Given that passing `None` for the filter seems to be confusing, I simply disabled it. We can re-evaluate in the future if there will be the need as you suggested.